### PR TITLE
Test strategy

### PR DESCRIPTION
**Pull request by 4ydan**
Looking for feedback to the test strategy and manual test template and protocol.

## Basics

- [x] I added a line to [/doc/CHANGELOG.md](/doc/CHANGELOG.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ x] The buildserver is happy.

## Checklist

- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [x] I created left-over issues for things that are still to be done
- [x] Code is conforming to [our Architecture](/doc/architecture)
- [x] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [x] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

File: doc/tests/README.md [hunk 1]
```diff
@@ -1,63 +1,73 @@
...
-- Business logic and data manipulation, e.g. code that performs calculations, manipulates data, or makes decisions based on input
-- Authentication and authorization
-- Error handling
+### 1. Unit Testing
 
-In the backend unit tests can be found in the `src/` directory.
-They are to be put in the same file as the functions or modules they are supposed to test.
-Create separate modules called test in that file annotated with `#[cfg(test)]` (that way they are only compiled when actually running tests).  
-See the [rust doc](https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html) for reference.
+Unit tests are used to test individual units of code in isolation from the rest of the system. This is important because it allows us to validate that each unit is working as intended, without interference from other parts of the system.
```
Comment (badnames) [inline doc/tests/README.md:40]:
```suggestion
Unit tests are used to test individual units of code in isolation from the rest of the system. 
This is important because it allows us to validate that each unit is working as intended, without interference from other parts of the system.
```
Our documentation guidelines state that there should be one sentence per line.

File: doc/tests/README.md [hunk 2]
```diff
@@ -70,15 +80,85 @@ In the application the integration tests will be used to test the following area
 In the backend integration tests can be found in the `src/test/` directory.
 The whole module is annotated with `#[cfg(test)]` and will therefore only be compiled for tests.
 
-## Manual Testing
+### 3. System Testing
 
-Manual testing will include performing an end-to-end testing of the application to ensure overall functionality and user experience.
+System testing will include performing an end-to-end testing of the application to ensure overall functionality and user experience.
```
Comment (badnames) [inline doc/tests/README.md:85]:
```suggestion
System testing will include performing and end-to-end testing of the application to ensure overall functionality and user experience.
```

**Review by badnames — APPROVED** [review]
Well done!

File: doc/tests/README.md (hunk 1 above)
Comment (4ydan) [reply doc/tests/README.md:40]: I missed that one, got it now, thank you.

File: doc/tests/README.md (hunk 2 above)
Comment (4ydan) [reply doc/tests/README.md:85]: it was jsut bad english from my side, I rewrote it to: "System testing will include end-to-end testing of the application to ensure overall functionality and user experience."

**Comment by 4ydan** [comment]
What should we do with:

### 4. Performance Testing

- ?

### 5. Security Testing

- ?

Do I leave it like this for now or just delete it from the doc?

File: doc/tests/protocols/template.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+# Test Protocol Template
+
+This is the raw template idea for executing a manual test.
```
Comment (markus2330) [inline doc/tests/protocols/template.md:3]:
```suggestion
This is the raw template for executing a manual test.
```

File: doc/tests/README.md [hunk 4]
```diff
@@ -70,15 +80,85 @@ In the application the integration tests will be used to test the following area
...
+Manual tests will be documented under `doc/tests/protocols`.
+
+- System tests must be protocolled and errors must be categorized by level and priority.
+- Test scenarios must be derived from functional requirements and use cases.
+- Testers must perform End-to-End testing of the software, including all user workflows and system integrations.
+- Exploratory testing techniques must be utilized to uncover potential issues and edge cases.
+
+### 4. Performance Testing
+
+- ?
```
Comment (markus2330) [inline doc/tests/README.md:99]: Should be written by @badnames (with some prior work by @kitzbergerg)

```suggestion
- TBD by @badnames
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Fascinating work, I love it! :heart:

**Comment by markus2330** [comment]
> Do I leave it like this for now or just delete it from the doc?

You can leave it now as suggested so that you don't need to reorder the numbers.

Can you also create follow-up issues for them?

> I couldnt find much in the doc about it. Maybe Christoph can help. I dont know his @.

@temmey

File: doc/CHANGELOG.md [hunk 5]
```diff
@@ -33,6 +33,7 @@ Syntax: `- short text describing the change _(Your Name)_`
 - Use the correct layer ID for create and fetch actions (for plantings) in the map editor, enabling the possibility for multiple independent maps _(Paul and Thorben)_
 - Location picker map in map creation form only loads after actively pressing a button _(Thorben)_
 - Maps can be duplicated from the map overview (in PR #515) _(Thorben)_
+- Test strategy written _(4ydan)_
```
[the remark this answers is not in the document]
Comment (4ydan) [reply doc/CHANGELOG.md:36]: oh yea thx

File: doc/tests/README.md [hunk 6]
```diff
@@ -1,63 +1,73 @@
...
 
-### Backend
+- End-to-End testing of critical workflows
+- Integration testing of various modules and components
+- System tests
+- Performance tests
+- Security tests
+- Regression testing
+- User acceptance testing
+- Compatibility testing on different platforms, browsers, and devices?
```
[the remark this answers is not in the document]
Comment (4ydan) [reply doc/tests/README.md:35]: Sadly no experience in that field.

File: doc/tests/README.md [hunk 7]
```diff
@@ -70,15 +80,85 @@ In the application the integration tests will be used to test the following area
...
+
+- Test environments must mirror production environments as closely as possible.
+  - Windows 10/11 or Linux with Python 11 and node v19 etc ...
+- Virtualization and containerization technologies, such as Docker and Kubernetes, must be utilized for test environment management.
+- Test environments must be version controlled and easily reproducible.
+
+## Testing Tools
+
+- Rust built-in tests by cargo
+- Selenium/Cucumber
```
[the remark this answers is not in the document]
Comment (4ydan) [reply doc/tests/README.md:115]: I will open a new issue for that.

File: doc/tests/README.md [hunk 8]
```diff
@@ -70,15 +80,85 @@ In the application the integration tests will be used to test the following area
...
+- Test data must cover a variety of scenarios, including at least valid and invalid cases.
+- Equivalence partitioning and boundary value analysis tests should be done when it makes sense.
+
+## Test Execution and Reporting
+
+- Test cases must be executed manually or via automation scripts, as appropriate for each testing type.
+- Unit tests must be executed before commiting to the VCS to speed up the detection of bugs and simple code breaks.
+- Integration tests must be executed by the CI pipeline after committing to the VCS or when opening a new PR.
+- Test metrics, such as test coverage and test execution time, must be tracked locally or in the automated pipeline.
+- System and User acceptance tests must be performed before every release.
```
[the remark this answers is not in the document]
Comment (4ydan) [reply doc/tests/README.md:130]: What exactly do you mean? Manual and acceptance tests before every PR? As for manual and user acceptance tests they will probably increase over time with added functionality. Performing and documenting manual tests will take over an hour if done properly. But it all depends on the scope of the tests.

**Comment by 4ydan** [comment]
Thank you very much for the feedback. I also added a naming convention for the protocols. If we are fine with that, this should be ready to merge and I will open two new issues regarding:
- Performance/Security testing strategies
- doc/decision for testing tool selection

File: doc/tests/protocols/README.md
Comment (4ydan) [inline doc/tests/protocols/README.md:1]: The naming convention.

File: doc/tests/README.md [hunk 9]
```diff
@@ -1,63 +1,76 @@
...
 
-    #[test]
-    fn test_plants() {
-      // Test complicated plants function here
-      todo!()
+        // Assert
+        assert_eq!(answer, 5);
     }
-}
 ```
```
[the remark this answers is not in the document]
Comment (4ydan) [reply doc/tests/README.md:63]: My bad, was not aware of that, thanks for pointing it out.

**Comment by markus2330** [comment]
Great work, thank you!

Btw. Would be great to also get a picture of you on the landing page :+1:  If you are concerned about privacy, it can be artistic.

File: doc/tests/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply doc/tests/README.md]: Yes, we need to find a good balance of automatic/manual tests to keep the overall time spent on a minimum. I agree some manual tests should be done on completion of every use case and before every release.
